### PR TITLE
Mirror heatmap controls

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Background
#### Link to issue 
https://biomage.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=BIOMAGE&modal=detail&selectedIssue=BIOMAGE-467
https://biomage.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=BIOMAGE&modal=detail&selectedIssue=BIOMAGE-626
#### Link to staging deployment URL 
https://ui-plsworkffs-ui149.scp-staging.biomage.net/
#### Anything else the reviewers should know about the changes here

# Changes
The selected genes from the heatmap are now stored in genes/expression/views/[plotUuid]
During developing I realised if we don't store the selected genes in the config anymore, they won't be persisted. For now the selected genes are also stored in the config so that we can save them.
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by ivababukova** [comment]
I just played around with it locally and I noticed that the heatmap in data exploration doesn't render properly - every time I add a gene, there are write lines all over the heatmap (see example screenshot). The problem could be that you are not making the heatmap enter loading state when we are adding/removing/overriding genes. Try adding the loading state back and this will probably get fixed.

[](url) 
Screenshot 2021-03-15 at 11 08 05

**Comment by StefanBabukov** [comment]
> I just played around with it locally and I noticed that the heatmap in data exploration doesn't render properly - every time I add a gene, there are write lines all over the heatmap (see example screenshot). The problem could be that you are not making the heatmap enter loading state when we are adding/removing/overriding genes. Try adding the loading state back and this will probably get fixed.
> 
> 
> Screenshot 2021-03-15 at 11 08 05

How do I reproduce it, I haven't been able to get those white lines.

**Comment by ivababukova** [comment]
I was just adding genes to the heatmap. But now I noticed that it appears on the master version of localhost, so maybe it is not from your PR. However, you do still need to trigger the heatmap into loading state when changing what genes to display.

File: src/components/ContentWrapper.jsx [hunk 1]
```diff
@@ -47,9 +47,10 @@ const ContentWrapper = (props) => {
   const pipelineStatusKey = pipelineStatus.pipeline?.status;
   const pipelineErrors = ['FAILED', 'TIMED_OUT', 'ABORTED'];
 
-  const pipelineRunning = pipelineStatusKey === 'RUNNING';
-  const pipelineRunningError = pipelineErrors.includes(pipelineStatusKey);
-
+  // const pipelineRunning = pipelineStatusKey === 'RUNNING';
+  // const pipelineRunningError = pipelineErrors.includes(pipelineStatusKey);
+  const pipelineRunning = false;
+  const pipelineRunningError = false;
```
Comment (ivababukova) [inline src/components/ContentWrapper.jsx]: @cosa65 do we need that merged in? Or is its purpose just to enable testing on staging?

Comment (StefanBabukov) [reply src/components/ContentWrapper.jsx]: I changed this temporarily so that staging can be opened I intended to revert it before merging.

Comment (ivababukova) [reply src/components/ContentWrapper.jsx]: sounds good, just make sure you don't forget about it

File: src/utils/plotSpecs/generateHeatmapSpec.js [hunk 2]
```diff
@@ -1,45 +1,45 @@
...
-        symbolSize: { value: 200 },
-        encode: {
-          labels: {
-            update: {
-              text: { scale: 'cellSetIDToName', field: 'label' },
-            },
+    },
+    {
+      fill: 'cellSetColors',
+      title: 'Cluster ID',
```
Comment (ivababukova) [inline src/utils/plotSpecs/generateHeatmapSpec.js]: why hardcoding the title here?

File: src/utils/plotSpecs/generateHeatmapSpec.js [hunk 3]
```diff
@@ -1,45 +1,45 @@
...
-        type: 'gradient',
-        orient: 'bottom',
-        direction: 'horizontal',
-        title: ['Intensity'],
-        labelFont: { value: config.fontStyle.font },
-        titleFont: { value: config.fontStyle.font },
-        gradientLength: {
-          signal: 'width',
-        },
+const generateSpec = (config) => {
```
Comment (ivababukova) [inline src/utils/plotSpecs/generateHeatmapSpec.js]: is this spec used for both heatmaps (data exploration and on plots and tables)?

File: src/redux/reducers/componentConfig/updateConfig.js [hunk 4]
```diff
@@ -4,7 +4,14 @@ const updateConfig = (state, action) => {
...
 
   const newConfig = _.cloneDeep(state[plotUuid].config);
-  _.merge(newConfig, configChange);
+
+  const mergeConfig = (objValue, srcValue) => {
+    if (_.isArray(objValue) && srcValue) {
+      return srcValue;
+    }
+  };
+
```
Comment (ivababukova) [inline src/redux/reducers/componentConfig/updateConfig.js:12]: why was this change necessary?

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 2 above)
Comment (StefanBabukov) [reply src/utils/plotSpecs/generateHeatmapSpec.js]: Because the legend displays the clusterIds that are shown (from metadata tracks) or does this need to be variable changed?
image

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 3 above)
Comment (StefanBabukov) [reply src/utils/plotSpecs/generateHeatmapSpec.js]: No the data-exploration heatmap spec is different (has tooltips, no second legend, mouseOver events, no config-set properties)

File: src/redux/reducers/componentConfig/updateConfig.js (hunk 4 above)
Comment (StefanBabukov) [reply src/redux/reducers/componentConfig/updateConfig.js:12]: Because before if you have a property e.g `something:['a',' b', 'c']` and you want to set it to ['a'] - the config change. The way that merge works (not an intersection) it won't overwrite the previous value of **something** to be equal to just ['a'] but instead it will remain unchanged. So this way we check if the value is an array and if the new value exists we set the previous one to the new one instead of doing a normal merge.

**Comment by cosa65** [comment]
There's merge conflicts.

File: src/components/plots/helpers/populateHeatmapData.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+import { union } from '../../../utils/cellSetOperations';
+
+const populateHeatmapData = (cellSets, config, expression, selectedGenes, downsampling = false) => {
+  const { hierarchy, properties, hidden } = cellSets;
+  const {
+    selectedTracks, groupedTracks, expressionValue,
+  } = config;
+  const maxCells = 1000;
+  const getCellsInSet = (cellSetName) => properties[cellSetName].cellIds;
+  // const selectedGenes = expression.views[componentType].data;
```
Comment (cosa65) [inline src/components/plots/helpers/populateHeatmapData.js]: Remove this

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 2 above)
Comment (cosa65) [inline src/utils/plotSpecs/generateHeatmapSpec.js]: We are using the heatmap in data integration with the titles like this: https://docs.google.com/presentation/d/1-6mtMrZF14GRAYtLjh3F7FmhyE4o5WRpwyxt5gISGDo/edit#slide=id.gae6b21872b_1_12

(with coloring by sample), before it was set up to allow for the title to be set according to what we need, is this causing problems on that side?

File: src/redux/reducers/componentConfig/updateConfig.js (hunk 4 above)
Comment (ivababukova) [reply src/redux/reducers/componentConfig/updateConfig.js:12]: I didn't understand your answer, can you give me a specific example of a case when that happens?

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 3 above)
Comment (ivababukova) [reply src/utils/plotSpecs/generateHeatmapSpec.js]: no, I meant whether this file "src/plotSpecs/generateHeatmapSpec.js" is used for generating both the data exploration and the plots and tables heatmap

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 2 above)
Comment (ivababukova) [reply src/utils/plotSpecs/generateHeatmapSpec.js]: It is always better to use a variable rather than hardocding that value. When you hardocde it like this, in the future if we wanted to display something else, not cluster id, we would have to change the config and adapt to it. Hardcoding values makes the code more fragile

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 3 above)
Comment (StefanBabukov) [reply src/utils/plotSpecs/generateHeatmapSpec.js]: It is only for plots and tables heatmap

File: src/redux/reducers/componentConfig/updateConfig.js (hunk 4 above)
Comment (StefanBabukov) [reply src/redux/reducers/componentConfig/updateConfig.js:12]: if we have a config field `{something: ['a', 'b','c']}` and we want to change it to `{something:['a']}` before it was `_.merge({something: ['a', 'b','c']}, {something:['a']}) ` - this would still return` {something: ['a', 'b','c']}`, not  `{something:['a']}`  due to how merge works https://lodash.com/docs/4.17.15  Example would be when you add and remove metadata tracks or selected genes - it was impossible to remove a selected gene before because the array would stay the same.

File: src/utils/plotSpecs/generateHeatmapSpec.js (hunk 2 above)
Comment (StefanBabukov) [reply src/utils/plotSpecs/generateHeatmapSpec.js:17]: I didn't get what you mean, I don't see a heatmap in the slides?

**Review by cosa65 — APPROVED** [review]

File: src/redux/reducers/componentConfig/updateConfig.js (hunk 4 above)
Comment (ivababukova) [reply src/redux/reducers/componentConfig/updateConfig.js:12]: thanks Stefan! I understand now

**Review by ivababukova — APPROVED** [review]
